File: ui-tests/PageObjects/Panes/Editor/FlowValue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects.Components;

namespace UiTests.PageObjects.Panes.Editor;

/// <summary>
/// Represents a single flow value rendered alongside the editor.
/// </summary>
public class FlowValue
{
    private static readonly string[] DefaultContextMenuEntries =
    {
        "Jump to value",
        "Add value to scratchpad",
        "Add all values to scratchpad"
    };

    private readonly ILocator _root;
    private readonly ContextMenu _contextMenu;

    public FlowValue(ILocator root, ContextMenu contextMenu)
    {
        _root = root ?? throw new ArgumentNullException(nameof(root));
        _contextMenu = contextMenu ?? throw new ArgumentNullException(nameof(contextMenu));
    }

    /// <summary>
    /// Root locator (the value box span element).
    /// </summary>
    public ILocator Root => _root;

    /// <summary>
    /// Checks if this flow value supports scratchpad operations.
    /// </summary>
    /// <remarks>
    /// Stdout flow values (with class flow-std-default-box) don't support
    /// scratchpad operations. They can be identified by their ID pattern:
    /// - Regular values: flow-{mode}-value-box-{i}-{step}-{expression}
    /// - Stdout values: flow-{mode}-value-box-{i}-{step} (no expression)
    /// </remarks>
    public async Task<bool> SupportsScratchpadAsync()
    {
        // Check if this is a stdout box which doesn't support scratchpad
        var classAttr = await _root.GetAttributeAsync("class");
        if (classAttr?.Contains("flow-std-default-box", StringComparison.OrdinalIgnoreCase) == true)
        {
            return false;
        }

        // Also check the ID pattern - regular values have expression name in the ID
        var id = await _root.GetAttributeAsync("id");
        if (string.IsNullOrEmpty(id))
        {
            return true; // Assume it supports scratchpad if no ID
        }

        // ID pattern: flow-{mode}-value-box-{i}-{step}-{expression}
        // Stdout pattern: flow-{mode}-value-box-{i}-{step}
        // Count hyphens: regular has at least 6 parts, stdout has only 5
        var parts = id.Split('-');
        return parts.Length >= 6;
    }

    /// <summary>
    /// Friendly name associated with the flow value (if rendered).
    /// </summary>
    public async Task<string> NameAsync()
    {
        var nameLocator = _root.Locator("xpath=preceding-sibling::span[contains(@class,'value-name')]" );
        if (await nameLocator.CountAsync() > 0)
        {
            var text = await nameLocator.Last.InnerTextAsync();
            return text?.TrimEnd(':').Trim() ?? string.Empty;
        }

        var fallback = await _root.GetAttributeAsync("data-expression");
        return fallback ?? string.Empty;
    }

    /// <summary>
    /// Textual representation of the value.
    /// </summary>
    public async Task<string> ValueTextAsync()
    {
        var text = await _root.InnerTextAsync();
        return text?.Trim() ?? string.Empty;
    }

    /// <summary>
    /// Returns the expected context menu entries for a flow value.
    /// </summary>
    public IReadOnlyList<string> ExpectedContextMenuEntries => DefaultContextMenuEntries;

    /// <summary>
    /// Opens the context menu for this flow value.
    /// </summary>
    /// <remarks>
    /// Flow values are rendered in Monaco content widgets which can interfere with
    /// standard Playwright right-click events. We use JavaScript to dispatch a
    /// contextmenu event directly to ensure reliable triggering.
    /// </remarks>
    public async Task<ContextMenu> OpenContextMenuAsync()
    {
        // First ensure the element is visible and scrolled into view
        await _root.ScrollIntoViewIfNeededAsync();

        // Get the element's bounding box for positioning
        var box = await _root.BoundingBoxAsync();
        if (box == null)
        {
            throw new InvalidOperationException("Flow value element is not visible or has no bounding box.");
        }

        // Dispatch contextmenu event via JavaScript for reliable triggering
        // in Monaco content widgets which may not receive standard click events
        await _root.EvaluateAsync(@"(element) => {
            const rect = element.getBoundingClientRect();
            const event = new MouseEvent('contextmenu', {
                bubbles: true,
                cancelable: true,
                view: window,
                button: 2,
                buttons: 2,
                clientX: rect.left + rect.width / 2,
                clientY: rect.top + rect.height / 2
            });
            element.dispatchEvent(event);
        }");

        await _contextMenu.WaitForVisibleAsync();
        return _contextMenu;
    }

    /// <summary>
    /// Reads the available context menu entries.
    /// </summary>
    public async Task<IReadOnlyList<string>> ContextMenuEntriesAsync()
    {
        var menu = await OpenContextMenuAsync();
        var entries = await menu.GetEntriesAsync();
        await menu.DismissAsync();
        return entries.Select(e => e.Text).ToList();
    }

    /// <summary>
    /// Selects a specific context menu option.
    /// </summary>
    public async Task SelectContextMenuOptionAsync(string option)
    {
        // For "Add value to scratchpad", use the more reliable CTRL+click method
        if (option.Contains("Add value to scratchpad", StringComparison.OrdinalIgnoreCase))
        {
            await AddToScratchpadAsync();
            return;
        }

        var menu = await OpenContextMenuAsync();
        await menu.SelectAsync(option);
    }

    /// <summary>
    /// Adds this flow value to the scratchpad using CTRL+click.
    /// </summary>
    /// <remarks>
    /// Flow values support CTRL+click as a shortcut to add to scratchpad.
    /// This is more reliable than using the context menu in Monaco content widgets.
    /// </remarks>
    public async Task AddToScratchpadAsync()
    {
        await _root.ScrollIntoViewIfNeededAsync();

        // Use CTRL+click which directly triggers the add-to-scratchpad action
        // without needing to open the context menu
        await _root.ClickAsync(new LocatorClickOptions
        {
            Modifiers = new[] { KeyboardModifier.Control }
        });
    }
}

File: tsc-ui-tests/page-objects/panes/editor/flow-value.ts
import type { Locator } from "@playwright/test";
import { ContextMenu } from "../../components/context-menu";

const DEFAULT_CONTEXT_MENU_ENTRIES = [
  "Jump to value",
  "Add value to scratchpad",
  "Add all values to scratchpad",
];

/**
 * Represents a single flow value rendered alongside the editor.
 *
 * Port of ui-tests/PageObjects/Panes/Editor/FlowValue.cs
 */
export class FlowValue {
  readonly root: Locator;
  private readonly contextMenu: ContextMenu;

  constructor(root: Locator, contextMenu: ContextMenu) {
    this.root = root;
    this.contextMenu = contextMenu;
  }

  /**
   * Checks if this flow value supports scratchpad operations.
   * Stdout flow values (with class flow-std-default-box) don't support scratchpad.
   */
  async supportsScratchpad(): Promise<boolean> {
    const classAttr = await this.root.getAttribute("class");
    if (classAttr?.includes("flow-std-default-box")) {
      return false;
    }

    const id = await this.root.getAttribute("id");
    if (!id) {
      return true;
    }

    // ID pattern: flow-{mode}-value-box-{i}-{step}-{expression}
    // Stdout pattern: flow-{mode}-value-box-{i}-{step}
    const parts = id.split("-");
    return parts.length >= 6;
  }

  /**
   * Friendly name associated with the flow value.
   */
  async name(): Promise<string> {
    const nameLocator = this.root.locator(
      "xpath=preceding-sibling::span[contains(@class,'value-name')]",
    );
    if ((await nameLocator.count()) > 0) {
      const text = await nameLocator.last().innerText();
      return text.replace(/:$/, "").trim();
    }

    const fallback = await this.root.getAttribute("data-expression");
    return fallback ?? "";
  }

  /**
   * Textual representation of the value.
   */
  async valueText(): Promise<string> {
    return (await this.root.innerText()).trim();
  }

  get expectedContextMenuEntries(): readonly string[] {
    return DEFAULT_CONTEXT_MENU_ENTRIES;
  }

  /**
   * Opens the context menu for this flow value.
   * Uses JavaScript event dispatch because Monaco content widgets
   * may not respond to standard Playwright right-click.
   */
  async openContextMenu(): Promise<ContextMenu> {
    await this.root.scrollIntoViewIfNeeded();

    await this.root.evaluate((element: Element) => {
      const rect = element.getBoundingClientRect();
      const event = new MouseEvent("contextmenu", {
        bubbles: true,
        cancelable: true,
        view: window,
        button: 2,
        buttons: 2,
        clientX: rect.left + rect.width / 2,
        clientY: rect.top + rect.height / 2,
      });
      element.dispatchEvent(event);
    });

    await this.contextMenu.waitForVisible();
    return this.contextMenu;
  }

  /**
   * Reads the available context menu entries.
   */
  async contextMenuEntries(): Promise<string[]> {
    const menu = await this.openContextMenu();
    const entries = await menu.getEntries();
    await menu.dismiss();
    return entries.map((e) => e.text);
  }

  /**
   * Adds this flow value to the scratchpad using Ctrl+click.
   */
  async addToScratchpad(): Promise<void> {
    await this.root.scrollIntoViewIfNeeded();
    await this.root.click({ modifiers: ["Control"] });
  }

  /**
   * Selects a specific context menu option.
   */
  async selectContextMenuOption(option: string): Promise<void> {
    if (option.toLowerCase().includes("add value to scratchpad")) {
      await this.addToScratchpad();
      return;
    }
    const menu = await this.openContextMenu();
    await menu.select(option);
  }
}
